docs/components/bucket.md

```diff
@@ -23,34 +23,37 @@ usage: thanos bucket [<flags>] <command> [<args> ...]
 inspect metric data in an object storage bucket
 
 Flags:
-  -h, --help                   Show context-sensitive help (also try --help-long
-                               and --help-man).
-      --version                Show application version.
-      --log.level=info         Log filtering level.
+  -h, --help                  Show context-sensitive help (also try --help-long
+                              and --help-man).
+      --version               Show application version.
+      --log.level=info        Log filtering level.
       --gcloudtrace.project=GCLOUDTRACE.PROJECT  
-                               GCP project to send Google Cloud Trace tracings
-                               to. If empty, tracing will be disabled.
+                              GCP project to send Google Cloud Trace tracings
+                              to. If empty, tracing will be disabled.
       --gcloudtrace.sample-factor=1  
-                               How often we send traces (1/<sample-factor>). If
-                               0 no trace will be sent periodically, unless
-                               forced by baggage item. See
-                               `pkg/tracing/tracing.go` for details.
-      --gcs-bucket=<bucket>    Google Cloud Storage bucket name for stored
-                               blocks.
-      --s3.bucket=<bucket>     S3-Compatible API bucket name for stored blocks.
-      --s3.endpoint=<api-url>  S3-Compatible API endpoint for stored blocks.
-      --s3.access-key=<key>    Access key for an S3-Compatible API.
-      --s3.insecure            Whether to use an insecure connection with an
-                               S3-Compatible API.
-      --s3.signature-version2  Whether to use S3 Signature Version 2; otherwise
-                               Signature Version 4 will be used.
-      --s3.encrypt-sse         Whether to use Server Side Encryption
-      --gcs-backup-bucket=<bucket>  
-                               Google Cloud Storage bucket name to backup blocks
-                               on repair operations.
-      --s3-backup-bucket=<bucket>  
-                               S3 bucket name to backup blocks on repair
-                               operations.
+                              How often we send traces (1/<sample-factor>). If 0
+                              no trace will be sent periodically, unless forced
+                              by baggage item. See `pkg/tracing/tracing.go` for
+                              details.
+      --provider.type=<provider>  
```

Comment:
All these parameters are S3 specific. :( I think best approach would be to specify provider config in external JSON file where you can specify type and all additional provider options, like Premium/Standard storage, access policies etc.. Each provider can have arbitrary options to pass there.

Reply:
Agree... We need some way to not exponentially increase number of flags, but also keep documentation clear, how to configure obj stores. I am generally against moving to a config file, instead of flags. This is basically harder to operate. There is no clear winner here..

Some quick idea:
- Put to flags only things that are common and add `objstore.extra` for file content and allow different providers parse it dynamically to internal structs?

Anything smarter? @brancz you are after inspiring Gophercon, maybe you have some idea (:

Reply:
I agree. I feel like we need another proposal for this as we need to consider various issues like performance and operating of yet another sidecar/service.

The problem is that this is quite critical part, but some gRPC S3 like API could make sense. It is not time critical though, as we have some safety buffer. I think I like this idea more & more. 

But clearly this change will not happen from day to day. I feel like we need something simple for time being...

Reply:
Added https://github.com/improbable-eng/thanos/issues/505